### PR TITLE
Allow multiple segments in Stringliterals

### PR DESCRIPTION
**Pull request by pawelsadlo**
This PR adds support for multi-segment literal strings.
for example
```scala
val path = root / "foo/bar" 
val qux = "qux"
val path = root / "foo/bar" / "baz" / qux
val path = root / "foo/bar" / "baz/qux"
```
it also parses `..` segments from literal as `os.up` enabling syntax like:
```scala
val path = root / "foo" / ".." / "bar" // equivalent to `root / "foo" / os.up / "bar"`
val path = root / "foo" /  "../bar" // equivalent to `root / "foo" / os.up / "bar"`
```
non-canonical paths used in literals result in compile-time errors, suggesting usage of canonical paths or removing path segment, eg.
```scala
val path = root / "foo/./bar" //suggests "foo/bar"
val path = root / "foo//bar" //suggests "foo/bar"
val path = root / "//foo//bar/./baz" //suggests "foo/bar/baz"

val path = root / ""  //suggests removing the segment
val path = root / "." //suggests removing the segment
val path = root / "/" //suggests removing the segment
val path = root / "//" //suggests removing the segment
val path = root / "/./" //suggests removing the segment

```

Note:
Its not usable in os-Lib itself, due to the fact that it would lead to macro expansion in the same compilation unit as its definition.

@lihaoyi 
there is a little bit of hacking involved:

1. There is a default implicit conversion not being a macro to be used within os library, without this we would have to add a submodule and split the whole project, I'm not even sure if its doable.
4. Needed to turn off acyclic in `Path` and particular `Macro` files (also needed to mock `acyclic.skipped` in case of `scala 3`).
5. Needed to provide another implicit conversion in `ViewBoundImplicit` trait because macros turn out to be not avaliable as implicit views, this is needed for `ArrayPathChunk` and `SeqPathChunk` to work.

### CODE REVIEW

**Comment by pawelsadlo** [comment]
@lihaoyi could you rerun checks?

**Comment by pawelsadlo** [comment]
@lihaoyi 
could you take a look at checks after [924a02e]? 
did you encounter something like this before?

`adding more tests[5a49ac2]` succeed in (ubuntu-latest, 17) and almost nothing changed since then (i have only added binary compatibility shims)

After last failed workflow I have changed macro to make StringPathChunk and ArrayPathChunk instead of SubPathChunk, but I dont know why SubPathChunk would fail..

**Comment by lihaoyi** [comment]
@pawelsadlo will take a look!

**Comment by pawelsadlo** [comment]
@lihaoyi it succeed now, so I dont know if we need to examine it, anyway could do a review?

File: os/src/Path.scala [hunk 1]
```diff
@@ -5,29 +5,52 @@ import java.nio.file.Paths
...
 }
-object PathChunk {
+trait StringPathChunkConversion {
+
+  implicit def stringToPathChunk(s: String): PathChunk =
+    new PathChunk.StringPathChunkInternal(s)
+}
+
+object PathChunk extends PathChunkMacros {
+  def segmentsFromString(s: String): Array[String] = {
```
Comment (pawelsadlo) [inline os/src/Path.scala:21]: @lihaoyi this is public only to make tests possible, however if you are fine with deleting tests for it, we can reduce accessibility

**Comment by lihaoyi** [comment]
Will take a look tomorrow

File: build.sc [hunk 2]
```diff
@@ -94,6 +95,13 @@ trait OsLibModule
 
 trait OsModule extends OsLibModule { outer =>
   def ivyDeps = Agg(Deps.geny)
+  override def compileIvyDeps = T{
+    scalaVersion.zip(super.compileIvyDeps).map{
+      case (scalaVer,superDeps) =>
+        val scalaReflectOpt = if (!ZincWorkerUtil.isDottyOrScala3(scalaVer)) Some(Deps.scalaReflect(scalaVer)) else None
+        superDeps ++ Agg.from(scalaReflectOpt)
+    }
+  }
```
Comment (lihaoyi) [inline build.sc:104]: This is probably better written as

```scala
  override def compileIvyDeps = T{
    val scalaReflectOpt = if (!ZincWorkerUtil.isDottyOrScala3(scalaVersion())) Some(Deps.scalaReflect(scalaVersion())) else None
    super.compileIvyDeps() ++ Agg.from(scalaReflectOpt)
  }
```

File: os/test/src/PathTests.scala [hunk 3]
```diff
@@ -2,16 +2,90 @@ package test.os
...
+
+      testSegmentsFromString("""ahs/""", "ahs" :: "" :: Nil)
+      testSegmentsFromString("""ahs//""", "ahs" :: "" :: "" :: Nil)
+
+      testSegmentsFromString("""ahs/aa/""", "ahs" :: "aa" :: "" :: Nil)
+      testSegmentsFromString("""ahs/aa//""", "ahs" :: "aa" :: "" :: "" :: Nil)
+
+      testSegmentsFromString("""/a""", "" :: "a" :: Nil)
+      testSegmentsFromString("""//a""", "" :: "" :: "a" :: Nil)
+      testSegmentsFromString("""//a/""", "" :: "" :: "a" :: "" :: Nil)
```
Comment (lihaoyi) [inline os/test/src/PathTests.scala:42]: Let's use the `List(...)` syntax here rather than `::` and `Nil`, for better consistency with the rest of the codebase

File: os/test/src/PathTests.scala [hunk 4]
```diff
@@ -2,16 +2,90 @@ package test.os
...
-
 import os._
-import os.Path.{driveRoot}
+import os.Path.driveRoot
+import os.PathChunk.segmentsFromString
 import utest.{assert => _, _}
+
 import java.net.URI
 object PathTests extends TestSuite {
+  private def nonValidPathSegment(chars: String) = s"[$chars] is not a valid path segment."
```
Comment (lihaoyi) [inline os/test/src/PathTests.scala:12]: I think we can do a more use-focused error message here, something like 

> literal path sequence <entire-literal> used in OS-Lib cannot have redundant slashes, please use <literal-without-leading-slashes> instead

File: os/test/src/PathTests.scala [hunk 5]
```diff
@@ -2,16 +2,90 @@ package test.os
...
+        compileError { """ root / "src/"  """ }.check("", nonValidPathSegment(""))
+        compileError { """ root / "src/." """ }.check("", nonValidPathSegment("."))
+        compileError { """ root / "src/.." """ }.check("", nonValidPathSegment(".."))
+
+        compileError { """ root / "" """ }.check("", nonValidPathSegment(""))
+        compileError { """ root / "." """ }.check("", nonValidPathSegment("."))
+        compileError { """ root / ".." """ }.check("", nonValidPathSegment(".."))
+
+        compileError(""" root / "hello" / ".." / "world" """).check("", nonValidPathSegment(".."))
+        compileError(""" root / "hello" / "../world" """).check("", nonValidPathSegment(".."))
```
Comment (lihaoyi) [inline os/test/src/PathTests.scala:81]: Should we allow these scenarios? Allowing literals to generate `ups` seems reasonable, given that they are literal strings, as long as they care canonicalized

**Comment by lihaoyi** [comment]
@pawelsadlo looks great, left some comments. If you update the code and update the PR description to have the most up to date state of the PR I think we can merge it

File: os/test/src/PathTests.scala (hunk 5 above)
Comment (pawelsadlo) [reply os/test/src/PathTests.scala:81]: I think it's ok, it gives lots of flexibility to users.

We should decide what to do with following edge cases:
`root / ".." / foo`
`root / "../.." / foo`
etc.
 
I think for the first one we should fail with compile error saying to use `up` instead of `".."`.
What do you think?

Comment (lihaoyi) [reply os/test/src/PathTests.scala:81]: Let's allow `root / ".." / foo`. It definitely duplicates `/ os.up /`, but both more concise and arguably more intuitive. And the compile-time-only nature of the conversions should stop the normal problems of untrusted user inputs causing filesystem traversal problems

File: os/test/src/PathTests.scala (hunk 4 above)
Comment (pawelsadlo) [reply os/test/src/PathTests.scala]: @lihaoyi
I have provided a `NonCanonicalLiteral` exception for this case
actually there is also an edge case when sanitizing literal results in empty path, for example `//`, `/.`, `.` literals , in such cases i throw:
```scala
InvalidSegment(
      literal,
      s"Literal path sequence [$literal] doesn't affect path being formed, please remove it"
    )
``` 
If you have any better suggestions for error messages I can change them.

File: os/test/src/PathTests.scala (hunk 3 above)
Comment (pawelsadlo) [reply os/test/src/PathTests.scala]: Actually I am thinking of making this function private[os] , in order not to expose it to public API. 

This means dropping tests, but keeping it public will force us to keep binary compatibility, and can result in users using it to form a paths (which we probably dont want)

Comment (lihaoyi) [reply os/test/src/PathTests.scala]: can we move the tests into the same package? I know the tests are mosty in a different package for now, but i dont think that's required

**Comment by pawelsadlo** [comment]
@lihaoyi could you please rerun checks?

**Comment by lihaoyi** [comment]
@pawelsadlo looks great, thanks! I merged it, I assume your bank details haven't changed so I'll pay out the bounty to the same account as last time
